### PR TITLE
Enabled rectilinear interpolation. 

### PR DESCRIPTION
**Pull request by jambo6**
Implemented functionality that enables `linear_interpolation_coeffs` to return the coefficients for a rectilinear interpolation. 

This involves a function `_prepare_rectilinear_interpolation` that converts a path of the form
```
[(t1, x1), (t2, NaN), (t3, x3), ...]
```
onto
```
[(t1, x1), (t2, x1), (t2, x1), (t3, x1), (t3, x3) ...]
```
such that linear interpolation on the modified path corresponds to rectilinear interpolation on the original path. 

This is invoked by supplying the index of the channel dimension where your time values belong (usually time_idx=0) as follows
```
    rectilinear_coeffs = linear_interpolation_coeffs(x, rectilinear=0)
```

### CODE REVIEW

File: torchcde/misc.py [hunk 1]
```diff
@@ -119,3 +119,34 @@ def validate_input_path(x, t):
...
+        out = x[np.arange(idx.shape[0])[:, None], idx]
+        return out
+
+    # Reshape to [N * C, L] and fill the 2d tensor
+    N, L, C = data.size()
+    data_shaped = data.transpose(1, 2).reshape(-1, L).numpy()
+    data_fill = fill2d(data_shaped).reshape(-1, C, L)
+    data_out = torch.Tensor(data_fill).transpose(1, 2)
+
+    return data_out
```
Comment (patrick-kidger) [inline torchcde/misc.py:152]: Untested but an alternative approach to forward-fill would be something like:
```
def forward_fill(data):
    nan_mask = torch.isnan(data)
    if nan_mask.any():
        data = data.clone()
        data_slices = iter(data.unbind(dim=-2))
        nan_mask_slices = iter(nan_mask.unbind(dim=-2))

        prev_data_slice = next(data_slices)
        next(nan_mask_slices)

        for data_slice, nan_mask_slice in zip(data_slices, nan_mask_slices):
            data_slice.masked_scatter_(nan_mask_slice, prev_data_slice.masked_select(nan_mask_slice))
            prev_data_slice = data_slice

    return data
```

This has a fast path for the common case of no NaNs, which is important. That fact aside, I don't know whether this will be meaningfull faster or slower than a torch-ified version of yours. Give it a test?

File: test/test_linear_interpolation.py [hunk 2]
```diff
@@ -114,3 +114,49 @@ def test_specification_and_derivative():
...
+        x_ffilled = ffill(x)
+        # Compute the true solution
+        N, L, C = x_ffilled.shape
+        rect_true = torch.zeros(N, 2 * L - 1, C)
+        lag = torch.cat([x_ffilled[:, 1:, [0]], x_ffilled[:, :-1, 1:]], -1)
+        rect_true[:, ::2, ] = x_ffilled
+        rect_true[:, 1::2] = lag
+        # Rectilinear solution
+        rectilinear = torchcde.interpolation_linear._prepare_rectilinear_interpolation(x, 0)
+        assert rect_true.allclose(rectilinear, equal_nan=True)
```
[the remark this answers is not in the document]
Comment (patrick-kidger) [reply test/test_linear_interpolation.py:162]: Also, is it really necessary to call the private function `_prepare...` directly? Can this not be tested via the public interface?

[the remark this answers is not in the document]
Comment (patrick-kidger) [reply test/test_linear_interpolation.py:162]: Additionally, there should be tests explicitly testing the edge cases of (a) a channel starting with nans; (b) a channel of all nans.

File: test/test_linear_interpolation.py [hunk 3]
```diff
@@ -114,3 +114,49 @@ def test_specification_and_derivative():
                         autoderivative = torch.stack(autoderivative).view(*evaluate.shape)
                         assert derivative.shape == autoderivative.shape
                         assert derivative.allclose(autoderivative, atol=1e-5, rtol=1e-5)
+
+
+def test_rectilinear_preparation():
+    # Some fucntions
+    pad_sequence = lambda sequences: torch.nn.utils.rnn.pad_sequence(sequences, batch_first=True, padding_value=nan)
```
[the remark this answers is not in the document]
Comment (jambo6) [reply test/test_linear_interpolation.py:121]: This has been amended. I have removed the intermediary function definitions.

File: test/test_linear_interpolation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jambo6) [reply test/test_linear_interpolation.py:162]: I have changed to go via linear_interpolation_coeffs. One thing I noticed though is this will backfill the data if a channel begins with nans. 

It would be good to have the option to fill those nans in with a zero value so as to not pass future information backwards.

File: torchcde/interpolation_linear.py [hunk 4]
```diff
@@ -82,7 +82,51 @@ def _linear_interpolation_coeffs_with_missing_values(t, x):
...
+        torchcde linear interpolation scheme to achieve rectilinear interpolation.
+
+    Returns:
+        A tensor, now of shape [N, 2*L - 1, C] that can be fed to linear interpolation coeffs to give rectilinear
+            coeffs.
+    """
+    # Check time_index is of the correct format
+    n_channels = data.size(2)
+    assert isinstance(time_index, int), "Index of the time channel must be an integer in [0, {}]".format(n_channels)
+    assert 0 <= time_index < data.size(2), "Time index must be in [0, {}], was given {}.".format(n_channels, time_index)
```
[the remark this answers is not in the document]
Comment (jambo6) [reply torchcde/interpolation_linear.py:112]: nice catch

File: torchcde/interpolation_linear.py [hunk 5]
```diff
@@ -82,7 +82,51 @@ def _linear_interpolation_coeffs_with_missing_values(t, x):
...
+    # Check the specified channel dim does not violate time conditions
+    times = data[:, :, time_index]
+    time_diffs = times[:, 1:] - times[:, :1]
+    assert time_diffs[~torch.isnan(time_diffs)].min() > 0, "Found consecutive times with a difference <= 0." \
```
[the remark this answers is not in the document]
Comment (jambo6) [reply torchcde/interpolation_linear.py:116]: I have added addtional checks to ensure no nan can occur before a time value, that is 
```
[2, nan, 1], [nan, 1, 2], [1, 2, nan, 3, nan]
```
will all now throw errors.

File: torchcde/misc.py (hunk 1 above)
Comment (jambo6) [reply torchcde/misc.py:152]: The performs a for loop over data_slices which is bound to be slower the inbuilt torch operations. I have rewritten forward fill so it is torch functions only and added the nan_mask.any() condition

File: torchcde/interpolation_linear.py [hunk 6]
```diff
@@ -82,7 +82,51 @@ def _linear_interpolation_coeffs_with_missing_values(t, x):
...
+            data_rect = [(t1, x1), (t2, x1), (t2, x1), (t3, x1), (t3, x3) ...]
+        This function simply performs the conversion from `data` to `data_rect` so that we can apply the inbuilt
+        torchcde linear interpolation scheme to achieve rectilinear interpolation.
+
+    Returns:
+        A tensor, now of shape [N, 2*L - 1, C] that can be fed to linear interpolation coeffs to give rectilinear
+            coeffs.
+    """
+    # Check time_index is of the correct format
+    n_channels = data.size(2)
```
[the remark this answers is not in the document]
Comment (jambo6) [reply torchcde/interpolation_linear.py:110]: Fixed and added tests.

File: test/test_linear_interpolation.py [hunk 7]
```diff
@@ -114,3 +114,49 @@ def test_specification_and_derivative():
...
+    for _ in range(5):
+        # Build some data with time
+        t_starts = torch.randn(5) ** 2
+        ts = [torch.linspace(t_start, t_start + 10, torch.randint(0, 50, (1,)).item()) for t_start in t_starts]
+        xs = [torch.randn(len(t), n_channels - 1) for t in ts]
+        x = pad_sequence([torch.cat([t_.view(-1, 1), x_], 1) for t_, x_ in zip(ts, xs)])
+        # Add some random nans about the place
+        mask = torch.randint(0, 5, (x.size(0), x.size(1), x.size(2) - 1)).to(float)
+        mask[mask == 0] = float('nan')
+        x[:, :, 1:] = x[:, :, 1:] * mask
```
[the remark this answers is not in the document]
Comment (jambo6) [reply test/test_linear_interpolation.py:151]: Have ensured an assertion is raised if time values are missin.

File: torchcde/interpolation_linear.py [hunk 8]
```diff
@@ -82,7 +82,51 @@ def _linear_interpolation_coeffs_with_missing_values(t, x):
...
+    times = data[:, :, time_index]
+    time_diffs = times[:, 1:] - times[:, :1]
+    assert time_diffs[~torch.isnan(time_diffs)].min() > 0, "Found consecutive times with a difference <= 0." \
+                                                           "Please ensure that data[:, :, 0] corresponds to times " \
+                                                           "and ensure that they are increasing for each sample."
+
+    # Forward fill and perform lag interleaving for rectilinear
+    data_filled = misc.torch_ffill(data)
+    data_repeat = data_filled.repeat_interleave(2, dim=1)
+    data_repeat[:, :-1, time_index] = data_repeat[:, 1:, time_index]
```
[the remark this answers is not in the document]
Comment (jambo6) [reply torchcde/interpolation_linear.py:123]: What error are you expecting? I have only verified it as far as my tests go.

File: test/test_linear_interpolation.py [hunk 9]
```diff
@@ -114,3 +115,71 @@ def test_specification_and_derivative():
...
+        rect_true[:, ::2, ] = x_ffilled
+        rect_true[:, 1::2] = lag
+        # Need to backfill rect true
+        # Rectilinear solution
+        rectilinear = torchcde.linear_interpolation_coeffs(x, rectilinear=0)
+        assert torch.equal(rect_true, rectilinear)
+
+
+if __name__ == '__main__':
+    test_rectilinear_preparation()
```
Comment (patrick-kidger) [inline test/test_linear_interpolation.py:185]: why is this here?

File: torchcde/interpolation_linear.py [hunk 10]
```diff
@@ -82,7 +82,58 @@ def _linear_interpolation_coeffs_with_missing_values(t, x):
...
+    # The following asserts ensure that no nan value is sandwiched in between time values
+    assert not torch.isnan(times[..., 0]).any(), "One or more of the time indexes start with a nan value."
+    nan_then_time_value = torch.isnan((times[..., :-1] * times[..., 1:])[~torch.isnan(times[..., 1:])]).any()
+    assert not nan_then_time_value, "One of more of the time paths has a nan in between non-nan times which is " \
+                                    "not allowed."
+    # This then asserts times are increasing
+    time_diffs = times[..., 1:] - times[..., :1]
+    assert time_diffs[~torch.isnan(time_diffs)].min() > 0, "Found consecutive times with a difference <= 0." \
+                                                           "Please ensure that data[:, :, time_idx] corresponds to " \
+                                                           "times and ensure that they are increasing for each sample."
```
Comment (patrick-kidger) [inline torchcde/interpolation_linear.py:125]: This logic seems rather involved + moderately expensive to evaluate. I think for now let's just disallow nan times. We can revisit this if we need to later.

File: torchcde/interpolation_linear.py [hunk 11]
```diff
@@ -91,6 +142,9 @@ def linear_interpolation_coeffs(x, t=None):
             length-many observations. Missing values are supported, and should be represented as NaNs.
         t: Optional one dimensional tensor of times. Must be monotonically increasing. If not passed will default to
             tensor([0., 1., ..., length - 1]).
+        rectilinear: Optional integer. Used for performing rectilinear interpolation. Default is None which results in
+            standard linear interpolation. For rectilinear interpolation time *must* be a channel in x and the
+            `rectilinear` parameter must be an integer specifying the channel index location of the time index in x.
```
Comment (patrick-kidger) [inline torchcde/interpolation_linear.py:147]: It would be nice to have a (brief one sentence) description of what rectilinear interpolation is here.

File: torchcde/misc.py [hunk 12]
```diff
@@ -119,3 +119,33 @@ def validate_input_path(x, t):
...
+    assert isinstance(data, torch.Tensor)
+    assert data.dim() >= 2
+
+    if torch.isnan(data).any():
+        # Reshape to [-1, L]
+        L, C = data.size()[-2:]
+        data_shaped = data.transpose(-1, -2).reshape(-1, L)
+        # Fill the 2d tensor along the length dim
+        mask = torch.isnan(data_shaped)
+        idx = torch.where(~mask, torch.arange(mask.shape[1]), 0)
```
Comment (patrick-kidger) [inline torchcde/misc.py:144]: I don't believe you've actually tested this. This will throw an error because 0 is not a Tensor.
The ~ is unnecessary because you can flip True/False cases.

Does the `arange` not need to have the device specified?

Stylisticly: prefer $tensor.size(x)$ over $tensor.shape[x]$.

File: torchcde/misc.py [hunk 13]
```diff
@@ -119,3 +119,33 @@ def validate_input_path(x, t):
...
+        cummax_idx, _ = torch.cummax(idx, dim=1)
+        data_fill = data_shaped[torch.arange(cummax_idx.shape[0])[:, None], cummax_idx].reshape(-1, C, L)
+        # Map back to original size
+        data_out = data_fill.transpose(-1, -2).reshape(data.size())
```
Comment (patrick-kidger) [inline torchcde/misc.py]: Is this really necessary? Can the whole operation not be done in the original format, without the overhead of two copies back and forth?

File: torchcde/misc.py [hunk 14]
```diff
@@ -119,3 +119,33 @@ def validate_input_path(x, t):
...
+    Returns:
+        tensor: Tensor with forward filled data.
+    """
+    # Checks
+    assert isinstance(data, torch.Tensor)
+    assert data.dim() >= 2
+
+    if torch.isnan(data).any():
+        # Reshape to [-1, L]
+        L, C = data.size()[-2:]
```
Comment (patrick-kidger) [inline torchcde/misc.py]: Stylisticly: prefer `data.shape[-2:]`

File: test/test_linear_interpolation.py [hunk 15]
```diff
@@ -114,3 +115,71 @@ def test_specification_and_derivative():
                         autoderivative = torch.stack(autoderivative).view(*evaluate.shape)
                         assert derivative.shape == autoderivative.shape
                         assert derivative.allclose(autoderivative, atol=1e-5, rtol=1e-5)
+
+
+def test_rectilinear_preparation(self):
```
[the remark this answers is not in the document]
Comment (jambo6) [reply test/test_linear_interpolation.py:120]: Whoops

File: torchcde/misc.py [hunk 16]
```diff
@@ -136,15 +138,17 @@ def forward_fill(x):
...
+        idx = torch.where(~mask, torch.arange(mask.shape[1]).to(mask.device), 0)
         cummax_idx, _ = torch.cummax(idx, dim=1)
-        x_fill = x_shaped[torch.arange(cummax_idx.shape[0])[:, None], cummax_idx].reshape(-1, C, L)
-        # Map back to original size
-        x = x_fill.transpose(-1, -2).reshape(x.size())
+        x_filled = x_shaped[torch.arange(cummax_idx.shape[0])[:, None], cummax_idx].reshape(x_transposed.size())
+        # Back to normal dim
+        x = x_filled.transpose(-1, fill_index)
 
     return x
```
Comment (patrick-kidger) [inline torchcde/misc.py:154]: Every comment previously made still applies.
Moreover it is once again clear that you have not actually tested this, as this code cannot run.

Try this instead:
```
def forward_fill(x, fill_index=-2):
    assert isinstance(x, torch.Tensor)
    assert x.dim() >= 2
    
    mask = torch.isnan(x)
    if mask.any():
        _, index = (~mask).cummax(dim=fill_index)
        x = x.gather(dim=fill_index, index=index)
        
    return x
```

I've not completely tested that but I think it's doing the right thing.

Bonus question for you: if using my version, what additional test should be added to ensure that things don't silently break in the future?

Comment (jambo6) [reply torchcde/misc.py:154]: So as not to tarnish my good name: It was tested, but turns out it fails for torch==1.6.0 but not torch==1.7.0. 

Neverthless, your code is far neater so I have replaced it. 

I have no idea regarding bonus questions, maybe testing cummax/gather to ensure they give the same outputs we expect now as the torch version changes?

Comment (patrick-kidger) [reply torchcde/misc.py:145]: Hah, that's fair!

And yup you're nearly there with the bonus question.

My code relies on the indices returned by cummax always being the last occurence of the maximum. For example, as specified by the documentation for `torch.cummax` at time of writing, it is ambiguous whether `torch.tensor([2, 2]).cummax().indices` should give 0s deterministically, 1s deterministically, or to randomly choose between them according to whatever rule it wants.

In this case I found out that `cummax` happens to use the desired behaviour just by testing it, but to be certain there should be a test for this, in case this changes in some later version of PyTorch. The test should be quite comprehensive: try tensors with different devices, different numbers of dimensions, which dimension the cummax is taken down, sizes of dimension (inc. important special cases: 0, 1, "small"~5, values either side of magic numbers like 32, "large"~1000); and indeed combinations of the above. It only needs to test bool-type tensors though as that's what's used here.

Incidentally if you want a nice example about not making assumptions for this kind of thing (at least using PyTorch 1.6.0 on Ubuntu):
```
import torch

def f(dtype, size):
    g = torch.Generator().manual_seed(456)
    print(torch.randn(size, dtype=dtype, generator=g)[0].item())

f(torch.float32, 15)
f(torch.float64, 15)
```
Prints out the same thing both times. We might reasonably assume that the choice of dtype doesn't matter. But then:
```
f(torch.float32, 16)
f(torch.float64, 16)
```
prints different things! It turns out there's a change in behaviour once the size of the tensor exceeds 16. Some other algorithm is used for which the choice of dtype now _does_ matter.

In fact not only does it print two different things, but even for the same dtype it's not consistent with the `size=15` case. It turns out that all `size<15` are the same as each other, and all `size>=16` are the same as each other, and that in the former case `float32` and `float64` just _happen_ to be the same as well.

Moreover note that I'm using `randn`. Using `rand` gives different results again.

Comment (jambo6) [reply torchcde/misc.py:145]: I do not think it is as complicated as this. We have a boolean tensor and simply wish cummax to give us the argument of the closest previous True value along a given dimension. With your 2 example the only implementation detail is that `torch.tensor([2, 2]).cummax().indices` could return `[0, 1], or [0, 0]`. It could never give `[1, 1]` since at the 0th argument it cannot return a larger cummax argument. 

Therefore the only check is that it will give `[0, 1]` not `[0, 0]`, as we want the most recent location. Thus I believe the test 
```
indices_check = torch.tensor([False, False, True, True]).to(device).cummax(0).indices
desired_indices = torch.tensor([0, 1, 2, 3]).to(device)
assert torch.equal(indices_check, desired_indices)
```
is more than sufficient. It simply ensures that if we have two Trues (i.e. two observed values), it will return to us the argument of the most recent true.

Comment (patrick-kidger) [reply torchcde/misc.py:145]: Cummax might do the correct thing on your particular test. But it might still do something else on a different input.

Anyway, relying on an implementation detail like this is pretty bad practice. Sorry, that's my bad - I shouldn't have suggested this as a solution in the first place.

Better (untested; going back to something like what you had):
```
    mask = torch.isnan(x)
    if mask.any():
        zero = torch.zeros((), device=x.device)
        arange = torch.arange(x.size(fill_index), device=x.device)
        for _ in range(x.dim() - fill_index if fill_index >= 0 else -fill_index):
            arange.unsqueeze_(-1)
        indices = torch.where(mask, zero, arange)
        _, index = indices.cummax(dim=fill_index)
        x = x.gather(dim=fill_index, index=index)
```

Comment (jambo6) [reply torchcde/misc.py:145]: I don't see how it will assuming it passes that test. 

Anyway, to avoid that altogether, I think this is a neater fix
```
    mask = torch.isnan(x)
    if mask.any():
        cumsum_mask = (~mask).cumsum(fill_index)
        cumsum_mask[mask] = 0
        _, index = (cumsum_mask).cummax(dim=fill_index)
        x = x.gather(dim=fill_index, index=index)
```

Comment (patrick-kidger) [reply torchcde/misc.py:147]: On the first point: you're assuming that it does the same thing when the tensor has 4 elements as when the tensor has 234 elements, for example. It need not.

Agreed, that is a neater fix. I like it.

**Comment by patrick-kidger** [comment]
And that's that! Congrats on your first pull request :)
